<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<HTML dir=ltr>
<HEAD>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=Windows-1252">
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-83731338-2"></script><script>  window.dataLayer = window.dataLayer || [];  function gtag(){dataLayer.push(arguments);}  gtag('js', new Date());  gtag('config', 'UA-83731338-2');</script><title>The try-except Statement</title>
<style>@import url(stylesheets/msdn_ie4.css);</style>
<link disabled rel="stylesheet" href="stylesheets/msdn_ie3.css"></HEAD>
<BODY>
<font face="verdana,arial,helvetica" size="2"><h1><a name="_clang_the_try.2d.except_statement"></a><sup></sup>The try-except Statement</h1>
<p>
<b>Microsoft Specific —&gt;</b></p>
<p>
The <b>try-except</b> statement is a Microsoft extension to the C language that enables applications to gain control of a program when events that normally terminate execution occur. Such events are called exceptions, and the mechanism that deals with exceptions is called structured exception handling. </p>
<p>
Exceptions can be either hardware- or software-based. Even when applications cannot completely recover from hardware or software exceptions, structured exception handling makes it possible to display error information and trap the internal state of the application to help diagnose the problem. This is especially useful for intermittent problems that cannot be reproduced easily. </p>
<p class=label>
<b>Syntax</b></p>
<p class=dt>
<i>try-except-statement</i> :</p>
<p class=indent>
<b>__try</b> <i>compound-statement</i><br>
<b>__except</b> <b>(</b> <i>expression</i> <b>)</b> <i>compound-statement</i></p>
<p>
The compound statement after the <b>__try</b> clause is the guarded section. The compound statement after the <b>__except</b> clause is the exception handler. The handler specifies a set of actions to be taken if an exception is raised during execution of the guarded section. Execution proceeds as follows:
<ol>
<li>
The guarded section is executed.<br><br></li>
<li>
If no exception occurs during execution of the guarded section, execution continues at the statement after the <b>__except</b> clause.<br><br></li>
<li>
If an exception occurs during execution of the guarded section or in any routine the guarded section calls, the<b>__except</b> expression is evaluated and the value returned determines how the exception is handled. There are three values:<p class=tl>
<b>EXCEPTION_CONTINUE_SEARCH</b>&nbsp;&nbsp;&nbsp;Exception is not recognized. Continue to search up the stack for a handler, first for containing <b>try-except</b> statements, then for handlers with the next highest precedence.</P><p class=tl>
<b>EXCEPTION_CONTINUE_EXECUTION</b>&nbsp;&nbsp;&nbsp;Exception is recognized but dismissed. Continue execution at the point where the exception occurred.</P><p class=tl>
<b>EXCEPTION_EXECUTE_HANDLER</b>&nbsp;&nbsp;&nbsp;Exception is recognized. Transfer control to the exception handler by executing the <b>__except</b> compound statement, then continue execution at the point the exception occurred.</P></li>
</ol>
<p>
Because the <b>__except</b> expression is evaluated as a C expression, it is limited to a single value, the conditional-expression operator, or the comma operator. If more extensive processing is required, the expression can call a routine that returns one of the three values listed above. </p>
<p class=indent>
<B><b>Note</b></B>&nbsp;&nbsp;&nbsp;Structured exception handling works with C and C++ source files. However, it is not specifically designed for C++. You can ensure that your code is more portable by using C++ exception handling. Also, the C++ exception handling mechanism is much more flexible, in that it can handle exceptions of any type.</p>
<p class=indent>
For C++ programs, C++ exception handling should be used instead of structured exception handling. For more information, see <a href="_pluslang_exception_handling.htm">Exception Handling</a> in the <i>C++ Language Reference</i>.</p>
<p>
Each routine in an application can have its own exception handler. The <b>__except</b> expression executes in the scope of the <b>__try</b> body. This means it has access to any local variables declared there.</p>
<p>
The <b>__leave</b> keyword is valid within a <b>try-except</b> statement block. The effect of <b>__leave</b> is to jump to the end of the <b>try-except</b> block. Execution resumes after the end of the exception handler. Although a <b>goto</b> statement can be used to accomplish the same result, a <b>goto</b> statement causes stack unwinding. The <b>__leave</b> statement is more efficient because it does not involve stack unwinding.</p>
<p>
Exiting a <b>try-except</b> statement using the <b>longjmp</b> run-time function is considered abnormal termination. It is illegal to jump into a <b>__try</b> statement, but legal to jump out of one. The exception handler is not called if a process is killed in the middle of executing a <b>try-except</b> statement.</p>
<p class=label>
<b>Example</b></p>
<p>
Following is an example of an exception handler and a termination handler. See <a href="_clang_the_try.2d.finally_statement.htm">The try-finally Statement</a> for more information about termination handlers. </p>
<pre><code>.
.
.
puts("hello");
__try{
   puts("in try");
   __try{
      puts("in try");
      RAISE_AN_EXCEPTION();
   }__finally{
      puts("in finally");
   }
}__except( puts("in filter"), EXCEPTION_EXECUTE_HANDLER ){
   puts("in except");
}
puts("world");
</code></pre>
<p>
This is the output from the example, with commentary added on the right:</p>
<pre><code>hello
in try&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /* fall into try&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; */
in try&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /* fall into nested try&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; */
in filter&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /* execute filter; returns 1 so accept&nbsp; */
in finally&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /* unwind nested finally&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; */
in except&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /* transfer control to selected handler */
world&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /* flow out of handler&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; */
</code></pre>
<p>
<b>END Microsoft Specific</b></p>
</font></BODY>
</HTML>
